tempodb/blocklist/poller.go

```diff
@@ -225,33 +267,84 @@ func (p *Poller) pollTenantAndCreateIndex(ctx context.Context, tenantID string)
 	return blocklist, compactedBlocklist, nil
 }
 
-func (p *Poller) pollTenantBlocks(ctx context.Context, tenantID string) ([]*backend.BlockMeta, []*backend.CompactedBlockMeta, error) {
-	blockIDs, err := p.reader.Blocks(ctx, tenantID)
+func (p *Poller) pollTenantBlocks(
+	ctx context.Context,
+	tenantID string,
+	previous backend.Blocklist,
+) ([]*backend.BlockMeta, []*backend.CompactedBlockMeta, error) {
+	currentBlockIDs, currentCompactedBlockIDs, err := p.reader.Blocks(ctx, tenantID)
 	if err != nil {
-		metricBlocklistErrors.WithLabelValues(tenantID).Inc()
-		return []*backend.BlockMeta{}, []*backend.CompactedBlockMeta{}, err
+		return nil, nil, err
 	}
 
+	previousMetas := previous.Metas(tenantID)
+	previousCompactedMetas := previous.CompactedMetas(tenantID)
+
 	bg := boundedwaitgroup.New(p.cfg.PollConcurrency)
-	chMeta := make(chan *backend.BlockMeta, len(blockIDs))
-	chCompactedMeta := make(chan *backend.CompactedBlockMeta, len(blockIDs))
+	chMeta := make(chan *backend.BlockMeta, len(currentBlockIDs))
```

Comment:
looking back, i really don't like that i used a channel as a buffer like this. would prefer just appending the block ids to a normal slice. it worked better when everything was async, now it's a bit weird.

not going to ask you to change it, but if you want to do a simpler pattern but felt bound to this, go for it.